pkg/util/push/push.go

```diff
@@ -17,14 +17,14 @@ func Handler(cfg distributor.Config, push func(context.Context, *client.WriteReq
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		compressionType := util.CompressionTypeFor(r.Header.Get("X-Prometheus-Remote-Write-Version"))
 		var req client.PreallocWriteRequest
-		req.Source = client.API
 		_, err := util.ParseProtoReader(r.Context(), r.Body, int(r.ContentLength), cfg.MaxRecvMsgSize, &req, compressionType)
 		logger := util.WithContext(r.Context(), util.Logger)
 		if err != nil {
 			level.Error(logger).Log("err", err.Error())
 			http.Error(w, err.Error(), http.StatusBadRequest)
 			return
 		}
+		req.Source = client.API
```

Comment:
why did this move?

Reply:
`_, err := util.ParseProtoReader(r.Context(), r.Body, int(r.ContentLength), cfg.MaxRecvMsgSize, &req, compressionType)` overwrite that value.

Reply:
I believe that was deliberate: if the caller supplies a value then we use that, otherwise we fall back to `API`.

Reply:
I tested here and I'm sure `util.ParseProtoReader` will overwrite `req.Source` to `0`.
Because `client.API` equals to `0`, you may not notice it.

**reproduce**
```diff
diff --git a/pkg/ingester/client/cortex.proto b/pkg/ingester/client/cortex.proto
index 6a341b59..a9dfcfee 100644
--- a/pkg/ingester/client/cortex.proto
+++ b/pkg/ingester/client/cortex.proto
@@ -30,7 +30,8 @@ service Ingester {
 message WriteRequest {
   repeated TimeSeries timeseries = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "PreallocTimeseries"];
   enum SourceEnum {
-    API = 0;
+    UNKNOWN_SOURCE = 0;
+    API = 2;
     RULE = 1;
   }
   SourceEnum Source = 2;
```
and print `req.Source` before and after`util.ParseProtoReader`.

result:
* before is `API`
* after is `UNKNOWN_SOURCE`

Reply:
Did you test it with and without the caller supplying a value in the protobuf?

Reply:
I just test it with real remote write from distributor.
I don't think remote write protobuf have `Source` field.

Reply:
remote write protobuf:
https://github.com/prometheus/prometheus/blob/master/prompb/remote.proto#L22-L24

Reply:
I agree on @Wing924 on this. The source is not part of the remote write protocol, so we need to inject it after the `WriteRequest` has been deseralized, otherwise the deserialization will overwrite it with the default value.

Reply:
I added test to cover both prometheus remote write and cortex writerequest.